Make action public

WiNloSt opened the pull request:
Epic: https://github.com/metabase/metabase/issues/27626

This PR adds a toggle to set an action public/non-public.

### Short demo
![toggle action public](https://user-images.githubusercontent.com/1937582/214648238-027ff29c-7871-4862-92e9-17bec99e9313.gif)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/27809)
<!-- Reviewable:end -->

WiNloSt committed:
Move action creator action buttons to the header following design

WiNloSt committed:
Remove double border which make it looks thicker

WiNloSt committed:
Draft toggle action public

kulyk commented:
@WiNloSt should we maybe extract `ActionCreator` refactoring into a different PR? We'd be able to review it quicker, and it'd be easier to follow changes when significant refactoring and feature work are not mixed together

WiNloSt committed:
Add confirmation when disabling public link similiar to questions and dashboards

WiNloSt committed:
Add an action public UUID input and copy button

WiNloSt commented:
@kulyk Ah yeah, haven't thought of that. That'd be easier to review. I'll create another PR and rebase this change of top of that one.

WiNloSt committed:
Show action settings based on user permission

WiNloSt committed:
Add public action toggle tests

WiNloSt commented on frontend/src/metabase/actions/containers/ActionCreator/InlineActionSettings.tsx:
```diff
@@ -0,0 +1,146 @@
+                <Icon name="info" size={10} />
+              </Tooltip>
+            </span>
+            <Toggle id={id} value={isPublic} onChange={handleTogglePublic} />
```
I deliberately use a different way of displaying the disable confirmation than in sharing questions and dashboards because they are rendering a different node so the animation on the toggle doesn't work.

https://github.com/metabase/metabase/blob/40bfc4a5543ba3f0fa1a23b97a4b0f6b5941ace9/frontend/src/metabase/public/components/widgets/SharingPane.tsx#L73-L99

#### toggle question/dashboard sharing
![toggle question public static](https://user-images.githubusercontent.com/1937582/214679048-9b6022a0-dcea-4ba2-9430-8210a498e71d.gif)

#### toggle action sharing
![toggle action public animation](https://user-images.githubusercontent.com/1937582/214679014-232d947e-9026-4499-b7c1-849674a3a907.gif)

WiNloSt commented on frontend/src/metabase/actions/containers/ActionCreator/InlineActionSettings.tsx:
```diff
@@ -0,0 +1,146 @@
+
+function isActionPublic(
+  action: WritebackAction,
+): action is PublicWritebackAction {
```
Need to use TypeScript's type predicate so it knows that if the check passes then `public_uuid` will not be null. I think there might be a better way than creating another type here like I imagine using something like `Required<WritebackAction, "public_uuid">` or something.

kulyk replied:
It's good enough IMO

WiNloSt commented on frontend/src/metabase/actions/containers/ActionCreator/InlineDataReference.styled.tsx:
```diff
@@ -1,16 +1,13 @@
   position: relative;
   height: 100%;
   background-color: ${color("white")};
-  border-left: 1px solid ${color("border")};
```
Remove double borders

#### Default view
<img width="612" alt="Screenshot 2023-01-20 at 11 18 00 AM" src="https://user-images.githubusercontent.com/1937582/214680210-287374a4-acf3-4f37-bed3-bebdc021521a.png">

#### Previous data reference
<img width="631" alt="Screenshot 2023-01-20 at 11 16 47 AM" src="https://user-images.githubusercontent.com/1937582/214679973-6ca8673e-c5e9-4a82-8e82-8857fd68b70c.png">
#### Current data reference (after fix)
<img width="610" alt="Screenshot 2023-01-20 at 11 17 19 AM" src="https://user-images.githubusercontent.com/1937582/214680169-527f18ac-cb8f-4821-b606-15a04509b2d8.png">

WiNloSt commented on frontend/src/metabase/actions/containers/ActionCreator/InlineDataReference.styled.tsx:
```diff
@@ -20,10 +17,3 @@ export const DataReferenceContainer = styled.div`
   }
 `;
-
-export const TriggerButton = styled(Button)`
```
This isn't needed anymore. Since we're rendering the trigger button inside `<ActionCreatorHeader />`

https://github.com/metabase/metabase/pull/27809/files#diff-a9d477e7189147f253f9e08ed333457267687b625a7eb4fb4db491bcf9ad23e5R19

WiNloSt commented on frontend/src/metabase/components/ConfirmContent.tsx:
```diff
@@ -6,7 +6,20 @@ import ModalContent from "metabase/components/ModalContent";
 
-const nop = () => {};
+const nop = () => {
+  // Do nothing
```
Without comment, ESLint would complain that we shouldn't write an empty function.

kulyk replied:
Should we use `_.noop`?

WiNloSt replied:
Good idea, so we don't need to declare `noop` everywhere. I didn't think of it much at first since this was a JS -> TS conversion and `nop` was already there. But I think I made another `noop` somewhere, and will remove that as well :)

WiNloSt commented on frontend/src/metabase/components/ConfirmContent.tsx:
```diff
@@ -6,7 +6,20 @@ import ModalContent from "metabase/components/ModalContent";
+  // Do nothing
+};
+
+interface ConfirmContentProps {
```
Converting to TypeScript is necessary because it's used in `<InlineActionSettings />` and TypeScript start inferring obscenely wrong types like `content: null` 🤦

WiNloSt commented on frontend/src/metabase/core/components/Button/Button.tsx:
```diff
@@ -127,7 +128,7 @@ const BaseButton = forwardRef(function BaseButton(
 });
 
-const Button = styled(BaseButton)`
+const Button = styled(BaseButton)<SpaceProps>`
```
After `<ConfirmContent />` has been converted to TypeScript it starts checking for type within that file that we use `ml` on Button so I need to add its props here.
https://github.com/metabase/metabase/blob/40bfc4a5543ba3f0fa1a23b97a4b0f6b5941ace9/frontend/src/metabase/components/ConfirmContent.tsx#L57

WiNloSt commented on frontend/src/metabase/lib/urls/actions.ts:
```diff
@@ -0,0 +1,3 @@
+export function publicAction(siteUrl: string, uuid: string) {
```
Uses a similar function like `publicQuestion` and `publicDashboard`

kulyk commented on frontend/src/metabase-types/api/mocks/actions.ts:
```diff
@@ -33,6 +33,7 @@ export const createMockQueryAction = ({
     parameters: [],
     created_at: new Date().toISOString(),
     updated_at: new Date().toISOString(),
+    public_uuid: opts.public_uuid ?? null,
```
Think we can just keep `null` there — `options` spreading should make sure the default value will be overwritten

WiNloSt replied:
Agree that's cleaner.

kulyk commented on frontend/src/metabase/entities/actions/actions.ts:
```diff
@@ -201,6 +206,50 @@ const Actions = createEntity({
   actions: {
     enableImplicitActionsForModel,
   },
+  objectActions: {
+    createPublicLink: createAction(
+      CREATE_PUBLIC_LINK,
+      ({ id }: { id: number }) => {
+        return ActionsApi.createPublicLink({ id }).then(
+          ({ uuid }: { uuid: string }) => {
+            return {
+              id,
+              uuid,
+            };
+          },
+        );
+      },
+    ),
+    deletePublicLink: createAction(
+      DELETE_PUBLIC_LINK,
+      ({ id }: { id: number }) => {
+        return ActionsApi.deletePublicLink({ id }).then(() => {
+          return {
+            id,
+          };
+        });
+      },
+    ),
+  },
```
Should we use `async/await` here?

WiNloSt replied:
TBH, I don't see much improvement from converting that to async/await here. In certain places, when dealing with lots of nested promise hell that could be a good reason to do async/await. But they're almost equal in my opinion here, I like using promise by default because when dealing with `catch` block it's easier and less verbose to write `.catch` than to write `try/catch` imo. What do you think?

ps. one interesting note is that VS code always prompt me to convert this to async/await, maybe it favor one over another 😃

kulyk commented on frontend/src/metabase/entities/actions/actions.ts:
```diff
@@ -201,6 +206,50 @@ const Actions = createEntity({
+  objectActions: {
+    createPublicLink: createAction(
+      CREATE_PUBLIC_LINK,
+      ({ id }: { id: number }) => {
```
Can we use `WritebackActionId` from `metabase-types/api` here?

WiNloSt replied:
Yeah, I could, but I wonder why we need to export every entity id type. And more confusingly some of them are `string` while some are `number`. In an ideal world, I'd expect every entity ID type to be the same e.g. string.

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/InlineActionSettings.tsx:
```diff
@@ -0,0 +1,146 @@
+
+interface StateProps {
+  siteUrl: string;
+  createPublicLink: ({ id }: { id: number }) => void;
+  deletePublicLink: ({ id }: { id: number }) => void;
```
Can we use `WritebackActionId` type here?

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/InlineActionSettings.tsx:
```diff
@@ -0,0 +1,146 @@
+  );
+};
+
+const mapStateToProps = (state: State) => ({
+  siteUrl: getSetting(state, "site-url"),
+});
+
+const mapDispatchToProps = {
+  createPublicLink: Actions.actions.createPublicLink,
+  deletePublicLink: Actions.actions.deletePublicLink,
+};
```
Think we prefer to keep `mapStateToProps` and `mapDispatchToProps` above the component?

WiNloSt replied:
You are right. A few quick searches have shown that that's the case.

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.tsx:
```diff
@@ -54,6 +58,8 @@ interface StateProps {
   actionId?: number;
   question?: Question;
   metadata: Metadata;
+  isAdmin: any;
```
`boolean`?

WiNloSt replied:
dang. Forgot to fix this. I put any there so I know I have to come back to fix it later, apparently forgot to annotate it with todo comment so I haven't fixed it in the end. Thanks for pointing this out.

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.tsx:
```diff
@@ -54,6 +58,8 @@ interface StateProps {
   actionId?: number;
```
Can we use `WritebackActionId` here please?

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -58,11 +77,24 @@ async function setup({ action }: SetupOpts = {}) {
   return { action };
 }
 
+const DATA_REFERENCE_ICON_LABEL = "reference icon";
+const ACTION_SETTINGS_ICON_LABEL = "gear icon";
```
Think it'd be better to use `getIcon` / `queryIcon` from `__support__/ui`:

```jsx
import { getIcon, queryIcon } from "__support__/ui"

expect(getIcon("reference")).toBeInTheDocument()
expect(queryIcon("gear")).not.toBeInTheDocument()
```

WiNloSt replied:
Didn't know these exists too.

kulyk replied:
They're pretty new!

WiNloSt replied:
unfortunately this doesn't work. The util return just the icon, but it's not clickable. I have a few options
1. keep using what's defined here
1. create another 2 util funtions i.e. `getIconButton` and `queryIconButton`
1. use `getIcon("reference").closest("button")` which kind of defeats the purpose of using a util function since we still need to manipulate it somehow.

I think 2. is ideal. What do you think?

kulyk replied:
Oh, I see; thanks for giving it a shot though. I don't have a solid opinion on that: some icon buttons are implemented as HTML buttons, and some use the `a` tag. A future with `getIcon`, `getIconButton`, and `getIconLink` makes me weep to be honest, so I'd stick to 1 🤷 Also, we can consider making `getIcon` / `queryIcon` a bit smarter by adding a `role` argument. What do you think about something like that?

```tsx
function getIcon(iconName: string, role?: string) {
  const name = `${iconName} icon`;
  return role
    ? screen.getByRole(role, { name })
    : screen.getByLabelText(name);
}
```

WiNloSt replied:
sounds good. I like that better than having 2 more similar implementations. I'll give it another shot, my first thought is that getByRole and getByLabelText is too different and imo shouldn't be in the same function, I think I can improve this a bit but gotta try.

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -35,19 +43,31 @@ function getTableObject(table: Table) {
   const scope = nock(location.origin);
 
   setupDatabasesEndpoints(scope, [getDatabaseObject(SAMPLE_DATABASE)]);
+  mockAdditionalEndpoints?.(scope, action);
```
Yeah, that's okay to mock less often-used endpoints in tests. Also, I think we should remove the `mockAdditionalEndpoints` argument and just make it always mock public link endpoints inside `setup`, WDYT?

WiNloSt replied:
Oh yeah, I can do that! I always thought nock will complain if we have unused mocks, so the merit of using `mockAdditionalEndpoints` is that you can pass mocks that will only be used in that specific test. From my testing, that isn't the case.

kulyk replied:
Nock should only complaint about calls to _not mocked_ endpoints, so it's okay to have more mocks than necessary. I doubt tests using `__support__/server-mocks` consume every single endpoint they spin up

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -35,19 +43,30 @@ function getTableObject(table: Table) {
 
 type SetupOpts = {
   action?: WritebackQueryAction;
+  storeInitialState?: Partial<State>;
```
Think the preferred way to setup state is to assemble its final shape inside `setup` and expose arguments to tweak important stuff, e.g.:

```jsx
function setup({ isAdmin, isSharingEnabled }) {
  const currentUser = createMockUser({ is_superuser: isAdmin })
  const settings = createMockSettingsState({ "enable-public-sharing": isSharingEnabled  })
  renderWithProviders(<JSX />, {
    storeInitialState: createMockState({ currentUser, settings })
  });
}
```

WiNloSt replied:
I could do that. It just feels like if we keep cramping more and more flags to the `setup` it's gonna be unbearable to read. That's why I chose to do this since I don't want to share much logic with other tests.

I'll try to put this logic inside `setup` and see how it becomes.

kulyk replied:
I think if it comes down to it, than it's a potential sign that we need to decompose the component

WiNloSt replied:
Doing this isn't too bad tbh. I've followed your suggestion.

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -90,6 +122,27 @@ describe("ActionCreator", () => {
       expect(screen.getByRole("button", { name: "Cancel" })).toBeEnabled();
     });
+
+    it("should show data reference icon and is clickable", async () => {
```
"is clickable" here sounds as if it's related to `it` rather than the data reference icon. Should it be something like "should show clickable data reference icon"?

kulyk commented:
@WiNloSt should we add an e2e test for enable/disable flow? There is a [test file](https://github.com/metabase/metabase/blob/master/frontend/test/metabase/scenarios/models/model-actions.cy.spec.js) for the actions tab already that can be extended

WiNloSt commented:
> @WiNloSt should we add an e2e test for enable/disable flow? There is a [test file](https://github.com/metabase/metabase/blob/master/frontend/test/metabase/scenarios/models/model-actions.cy.spec.js) for the actions tab already that can be extended

@kulyk Oh didn't realize we have that file. I'll add some basic scenarios there. Good call 👍

WiNloSt committed:
Remove unused import 🤦‍

WiNloSt commented on frontend/src/metabase-types/api/mocks/actions.ts:
```diff
@@ -33,6 +33,7 @@ export const createMockQueryAction = ({
     parameters: [],
     created_at: new Date().toISOString(),
     updated_at: new Date().toISOString(),
+    public_uuid: null,
```
Update these mock functions to ensure `public_uuid` property will always exist.

WiNloSt commented on frontend/test/__support__/ui.tsx:
```diff
@@ -154,12 +155,12 @@ function MaybeDNDProvider({
-export function getIcon(name: string) {
-  return screen.getByLabelText(`${name} icon`);
+export function getIcon(name: ByRoleMatcher, role: ByRoleMatcher = "img") {
+  return screen.getByRole(role, { name: `${name} icon` });
```
@kulyk Worth pointing out that I swap the implementation to `*ByRole`, and all tests still pass (at least locally need to wait and see the CI build too). So I guess that's better since we can target icon buttons now.

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -120,5 +163,104 @@ describe("ActionCreator", () => {
+          screen.queryByRole("textbox", { name: "Public action link URL" }),
+        ).not.toBeInTheDocument();
+
+        // Make the action public
+        screen.getByRole("switch", { name: "Make public" }).click();
```
I think the `name` here makes it clear what's going on; should we remove the comment?

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -120,5 +163,104 @@ describe("ActionCreator", () => {
+        // Make the action public
+        screen.getByRole("switch", { name: "Make public" }).click();
+
+        await waitFor(() => {
+          expect(
+            screen.getByRole("switch", { name: "Make public" }),
+          ).toBeChecked();
+        });
```
Let's use `await screen.findByRole` instead of `waitFor` here

WiNloSt replied:
afair it doesn't work but I'll give it another try. `await screen.findByRole` will get you an element. And since the elements is already on screen the test will fail since we haven't waited for it to turn unchecked.

From my experience, uses `await screen.findBy` to get element that isn't on screen yet. And use `waitFor` if you expects existing elements to change in another way.

WiNloSt replied:
This confirms it doesn't work.
<img width="1361" alt="image" src="https://user-images.githubusercontent.com/1937582/215116098-545449f2-5390-423c-b383-6895c2bc88e7.png">

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -120,5 +163,104 @@ describe("ActionCreator", () => {
+          screen.getByRole("textbox", { name: "Public action link URL" }),
+        ).toHaveValue(expectedPublicLinkUrl);
+
+        // Disable public sharing
+        screen.getByRole("switch", { name: "Make public" }).click();
+        expect(
+          screen.getByRole("heading", { name: "Disable this public link?" }),
+        ).toBeInTheDocument();
+        screen.getByRole("button", { name: "Yes" }).click();
```
The code is pretty self-explanatory IMO, I think we can remove the comment

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -120,5 +163,104 @@ describe("ActionCreator", () => {
+        ).toBeInTheDocument();
+        screen.getByRole("button", { name: "Yes" }).click();
+
+        await waitFor(() => {
+          expect(
+            screen.getByRole("switch", { name: "Make public" }),
+          ).not.toBeChecked();
+        });
```
Let's use `await screen.findByRole` instead of `waitFor`

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreatorView.tsx:
```diff
@@ -19,16 +18,22 @@ import {
+  hasSharingPermission: boolean;
   canSave: boolean;
 
+  actionId?: any;
```
`WritebackActionId`?

WiNloSt replied:
🤦 I missed this again.

kulyk commented on frontend/src/metabase/entities/actions/actions.ts:
```diff
@@ -176,6 +179,9 @@ const enableImplicitActionsForModel =
     dispatch(Actions.actions.invalidateLists());
   };
 
+const CREATE_PUBLIC_LINK = "metabase/action/CREATE_PUBLIC_LINK";
+const DELETE_PUBLIC_LINK = "metabase/action/DELETE_PUBLIC_LINK";
```
If we're implementing this as an entity action, I think we should follow the naming pattern here:

```
metabase/entities/actions/CREATE_PUBLIC_LINK
metabase/entities/actions/DELETE_PUBLIC_LINK
```

?

WiNloSt replied:
Oh yeah, totally forgot about that. In the first version, I created these actions outside the entity like for public dashboards and questions.

https://github.com/metabase/metabase/blob/2ee1479da12244f84f402b1c0d93bd1df811f191/frontend/src/metabase/dashboard/actions/sharing.js#L35

kulyk commented on frontend/src/metabase/entities/actions/actions.ts:
```diff
@@ -201,6 +207,50 @@ const Actions = createEntity({
+    createPublicLink: createAction(
+      CREATE_PUBLIC_LINK,
+      ({ id }: { id: WritebackActionId }) => {
+        return ActionsApi.createPublicLink({ id }).then(
+          ({ uuid }: { uuid: string }) => {
+            return {
+              id,
+              uuid,
+            };
+          },
+        );
```
Do you think it'd be possible to do an API call and then dispatch `Actions.actionTypes.update`, so we don't need to extend the entity reducer?

WiNloSt replied:
I was trying to do that, but didn't know how to do it.

WiNloSt replied:
@kulyk Seems `Entity.actionTypes.UPDATE` expects the payload to be normalized by `normalizr` which we don't necessary have to do in this case. It would be indirect in my opinion to try to reuse that action since we need to try to construct a payload that's normalized by `normalizr` manually which doesn't sound too enticing to me.
<img width="724" alt="image" src="https://user-images.githubusercontent.com/1937582/215120705-7d0a636f-b5a1-4b1a-8c1d-90e7f9e22aad.png">

I honestly expect we expose some kind of action that would update the entity automatically by passing it partial properties that are changed, but `*.actions.update` is currently reserved for calling the endpoint that seem there's no way to tell the action "Hey I just want to update the store" yet.

https://github.com/metabase/metabase/blob/a243b16cca8b425c303961d95429752997db4641/frontend/src/metabase/lib/entities.js#L257-L259

I feel like this would require more tinkering, what do you think about leaving it as is for now?

kulyk replied:
Yeah, let's keep things as is if it's not trivial

kulyk commented on frontend/src/types/utils.d.ts:
```diff
@@ -0,0 +1,3 @@
+type WithRequired<T, K extends keyof T> = T & {
+  [P in K]-?: T[K];
+};
```
Not sure I can find if it's used anywhere?

WiNloSt replied:
dang. that's just something I played with, forgot to remove it out. 🙏

kulyk commented on frontend/src/metabase/entities/actions/actions.ts:
```diff
@@ -176,6 +179,9 @@ const enableImplicitActionsForModel =
     dispatch(Actions.actions.invalidateLists());
   };
 
+const CREATE_PUBLIC_LINK = "metabase/entities/action/CREATE_PUBLIC_LINK";
+const DELETE_PUBLIC_LINK = "metabase/entities/action/DELETE_PUBLIC_LINK";
+
```
@WiNloSt entity actions should use plural naming, i.e. `/actions/`, not `/action/`

WiNloSt replied:
Dang it! Why do I miss this over and over again.

<img src="https://media0.giphy.com/media/3ohs7KViF6rA4aan5u/giphy.gif?cid=ecf05e47l8z8zhjyz60s9p5fbly3yi363hcb0bh9a6p04ovl&rid=giphy.gif&ct=g" />

kulyk reviewed (approved):
Great work! Just a minor little comment

WiNloSt commented on frontend/src/metabase/entities/actions/actions.ts:
```diff
@@ -176,6 +179,9 @@ const enableImplicitActionsForModel =
     dispatch(Actions.actions.invalidateLists());
   };
 
+const CREATE_PUBLIC_LINK = "metabase/entities/actions/CREATE_PUBLIC_LINK";
+const DELETE_PUBLIC_LINK = "metabase/entities/actions/DELETE_PUBLIC_LINK";
```
@kulyk Hope, I got it right this time! 😬

ranquild commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -120,5 +163,102 @@ describe("ActionCreator", () => {
+          isPublicSharingEnabled: true,
+        });
+
+        getIcon("gear", "button").click();
```
AFAIK it should be `userEvent.click(getIcon(...))`

kulyk commented on frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.unit.spec.tsx:
```diff
@@ -120,5 +163,102 @@ describe("ActionCreator", () => {
+        expect(getIcon("gear", "button")).toBeInTheDocument();
+      });
+
+      it("should be able to enabled action public sharing", async () => {
```
Typo:

```suggestion
      it("should be able to enable action public sharing", async () => {
```

WiNloSt committed:
Attempt to fix flaky CI unit tests

The problem seems to be because of how long it takes for the response
to be received even on unit tests, it took longer than 1 second which is
the default timeout for `waitFor`.

WiNloSt commented on frontend/test/jest-setup.js:
```diff
@@ -1,4 +1,3 @@
-import { XMLHttpRequest } from "xmlhttprequest";
```
We added this when introducing `nock` in https://github.com/metabase/metabase/pull/27337. It seems we don't really need it.

WiNloSt committed:
See if not using `userEvent` could make the involving nock faster

WiNloSt committed:
Improve test speed to reduce flakiness
